src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/SubqueryIT.java

```diff
@@ -267,4 +274,101 @@ public void nonCorrelatedNotExistsParentWhere() throws IOException {
+
+        assertThat(getTotalHits(result), equalTo(1000));
+        JSONObject gender = (JSONObject) result.query("/aggregations/gender");
+        assertThat(gender.getJSONArray("buckets").length(), equalTo(2));
+
+        boolean isMaleFirst = gender.optQuery("/buckets/0/key").equals("m");
+        int maleBucketId = isMaleFirst ? 0 : 1;
+        int femaleBucketId = isMaleFirst ? 1 : 0;
+
+        String maleBucketPrefix = String.format(Locale.ROOT, "/buckets/%d", maleBucketId);
+        String femaleBucketPrefix = String.format(Locale.ROOT, "/buckets/%d", femaleBucketId);
+
+        assertThat(gender.query(maleBucketPrefix + "/key"), equalTo("m"));
+        assertThat(gender.query(maleBucketPrefix + "/count/value"), equalTo(507));
+        assertThat(gender.query(femaleBucketPrefix + "/key"), equalTo("f"));
+        assertThat(gender.query(femaleBucketPrefix + "/count/value"), equalTo(493));
+    }
+
+    @Test
+    public void selectFromSubqueryWithCountAndGroupByAndOrderByShouldPass() throws IOException {
+        JSONObject result = executeQuery(
+                String.format(
+                        "SELECT t.TEMP as count " +
+                        "FROM (SELECT COUNT(*) as TEMP FROM %s/account GROUP BY age ORDER BY TEMP) t",
+                        TEST_INDEX_ACCOUNT));
+        JSONArray buckets = (JSONArray) result.query("/aggregations/age/buckets");
+        List<Integer> countList = new ArrayList<>();
+        for (int i = 0; i < buckets.length(); ++i) {
+            countList.add((int) buckets.query(String.format(Locale.ROOT, "/%d/count/value", i)));
+        }
+
+        assertThat(countList, isInOrder(true));
+    }
+
+    @Test
+    public void selectFromSubqueryWithCountAndGroupByAndHavingShouldPass() throws Exception {
+        JSONObject result = executeQuery(
+                String.format("SELECT t.T1 as g, t.T2 as c " +
+                              "FROM (SELECT gender as T1, COUNT(*) as T2 " +
+                              "      FROM %s/account " +
+                              "      GROUP BY gender " +
+                              "      HAVING T2 > 500) t", TEST_INDEX_ACCOUNT));
+        assertThat(result.query("/aggregations/g/buckets/0/c/value"), equalTo(507));
+    }
+
+    @Test
+    public void selectFromSubqueryCountAndSum() throws IOException {
+        JSONObject result = executeQuery(
+                String.format(
+                        "SELECT t.TEMP1 as count, t.TEMP2 as balance " +
+                        "FROM (SELECT COUNT(*) as TEMP1, SUM(balance) as TEMP2 " +
+                        "      FROM %s/account) t",
+                        TEST_INDEX_ACCOUNT));
+
+        assertThat(result.query("/aggregations/count/value"), equalTo(1000));
+        assertThat(result.query("/aggregations/balance/value"), equalTo(25714837.0));
+    }
+
+    private Matcher<? super List<? extends Comparable>> isInOrder(boolean asc) {
```

Comment:
That's useful. Could you move it to MatcherUtils?

Reply:
I perfer to use `assertTrue(Ordering.natural().isOrdered(countList));` instead of add another matcher.
I think maybe this is the reason why this request has been in pending status for long time. https://code.google.com/archive/p/hamcrest/issues/174